File: src/main/java/com/mkreidl/ephemeris/sky/RiseSetCalculator.java
package com.mkreidl.ephemeris.sky;

import com.mkreidl.ephemeris.Time;
import com.mkreidl.ephemeris.geometry.Angle;
import com.mkreidl.ephemeris.geometry.Circle;
import com.mkreidl.ephemeris.geometry.Spherical;
import com.mkreidl.ephemeris.geometry.Stereographic;
import com.mkreidl.ephemeris.sky.coordinates.Equatorial;
import com.mkreidl.ephemeris.solarsystem.Body;

public abstract class RiseSetCalculator
{
    public enum EventType
    {
        RISE( -1 ), TRANSIT( 0 ), SET( 1 );
        private final int signum;

        EventType( int signum )
        {
            this.signum = signum;
        }
    }

    public enum LookupDirection
    {
        FORWARD, BACKWARD
    }

    public static RiseSetCalculator getDummy()
    {
        return new RiseSetCalculator()
        {
            @Override
            public boolean compute( long startTimeMs )
            {
                return false;
            }
        };
    }

    public static final double OPTICAL_HORIZON_DEG = -34.0 / 60;
    private static final double RAD_TO_SIDEREAL_MILLIS = Time.MILLIS_PER_SIDEREAL_DAY / ( 2 * Math.PI );

    protected final Equatorial.Sphe topocentric = new Equatorial.Sphe();
    protected final Spherical geographicLocation = new Spherical( Body.EARTH.RADIUS_MEAN_M, 0, 0 );
    protected final Time time = new Time();
    protected EventType mode = EventType.SET;

    LookupDirection lookupDirection = LookupDirection.FORWARD;
    double virtualHorizonDeg = OPTICAL_HORIZON_DEG;

    private final Circle horizon = new Circle();
    private Stereographic projection;
    private long startTimeMs;

    public abstract boolean compute( long startTimeMs );

    public long getTime()
    {
        return time.getTime();
    }

    private double computeHourAngle()
    {
        // Right ascension is the sidereal time at (upper) meridian transit (hourAngle == 0)
        final double siderealTime = time.getHourAngleOfVernalEquinox() + geographicLocation.lon;
        return Angle.standardize( siderealTime - topocentric.lon );
    }

    private double getOrbitRadius()
    {
        final double z = Math.sin( geographicLocation.lat > 0 ? topocentric.lat : -topocentric.lat );
        return Math.sqrt( ( 1 + z ) / ( 1 - z ) );
    }

    public void setEventType( EventType mode )
    {
        this.mode = mode;
    }

    public void setSearchDirection( LookupDirection lookupDirection )
    {
        this.lookupDirection = lookupDirection;
    }

    public void setGeographicLocation( Spherical geographicLocation )
    {
        setGeographicLocation( geographicLocation.lon, geographicLocation.lat );
    }

    public void setGeographicLocation( double lon, double lat )
    {
        geographicLocation.lon = lon;
        geographicLocation.lat = lat;
        projection = new Stereographic( geographicLocation.lat >= 0 ? 1 : -1 );
        updateHorizon();
    }

    void setStartTime( long startTimeMs )
    {
        this.startTimeMs = startTimeMs;
        time.setTime( startTimeMs );
    }

    boolean isCrossing()
    {
        return mode == EventType.TRANSIT || !completelyAboveHorizon() && !completelyBelowHorizon();
    }

    private boolean completelyAboveHorizon()
    {
        return Math.abs( Math.toDegrees( geographicLocation.lat + topocentric.lat ) ) >= 90 + virtualHorizonDeg;
    }

    private boolean completelyBelowHorizon()
    {
        return Math.abs( Math.toDegrees( geographicLocation.lat - topocentric.lat ) ) >= 90 - virtualHorizonDeg;
    }

    private void updateHorizon()
    {
        projection.project( geographicLocation, Math.toRadians( 90 - virtualHorizonDeg ), horizon );
    }

    boolean adjustTime()
    {
        updateHorizon();
        final double alpha;
        switch ( mode )
        {
            case RISE:
            case SET:
                alpha = mode.signum * computeHourAngleAtSet() - computeHourAngle();
                break;
            default:
                alpha = -computeHourAngle();
                break;
        }
        time.addMillis( (long)( alpha * RAD_TO_SIDEREAL_MILLIS ) );
        if ( lookupDirection == LookupDirection.FORWARD && time.getTime() < startTimeMs )
            time.addMillis( Time.MILLIS_PER_SIDEREAL_DAY );
        if ( lookupDirection == LookupDirection.BACKWARD && time.getTime() > startTimeMs )
            time.addMillis( -Time.MILLIS_PER_SIDEREAL_DAY );
        return !Double.isNaN( alpha ) && !Double.isInfinite( alpha );
    }

    private double computeHourAngleAtSet()
    {
        final double dist = horizon.distFromOrigin();
        final double rHor = horizon.r;
        final double rOrb = getOrbitRadius();
        return Math.acos( ( rHor * rHor - dist * dist - rOrb * rOrb ) / ( 2 * dist * rOrb ) );
    }
}

File: src/main/java/com/mkreidl/ephemeris/sky/RiseSetCalculator.kt
package com.mkreidl.ephemeris.sky

import com.mkreidl.ephemeris.MILLIS_PER_SIDEREAL_DAY
import com.mkreidl.ephemeris.SIDEREAL_MILLIS_PER_RADIAN
import com.mkreidl.ephemeris.time.Instant
import com.mkreidl.math.Angle
import com.mkreidl.math.Circle
import com.mkreidl.math.Spherical3
import com.mkreidl.math.Stereographic

abstract class RiseSetCalculator(protected val mode: EventType, protected val lookupDirection: LookupDirection) {

    enum class EventType(internal val sign: Double) {
        RISE(-1.0),
        TRANSIT(0.0),
        SET(1.0)
    }

    enum class LookupDirection(internal val sign: Int) {
        FORWARD(1),
        BACKWARD(-1)
    }

    abstract fun compute(): Boolean

    protected var topos = Topos(0.0, 0.0, Instant.J2000)
    private var projection = Stereographic.N
    private lateinit var horizon: Circle

    var geographicLocation = Spherical3.ZERO
        set(location) {
            field = location
            topos = topos.copy(longitude = location.lon, latitude = location.lat)
            projection = if (geographicLocation.lat >= 0) Stereographic.N else Stereographic.S
        }

    var time = Instant.J2000
        protected set(time) {
            field = time
            topos = topos.copy(instant = time)
        }

    protected var startTime = Instant.J2000
        set(startTime) {
            field = startTime
            time = startTime
        }

    fun setStartTimeEpochMilli(epochMilli: Long) {
        startTime = Instant.ofEpochMilli(epochMilli)
        time = startTime
    }

    protected var virtualHorizon = OPTICAL_HORIZON
    protected var topocentric = Spherical3.ZERO

    protected fun adjustTime(): Boolean {
        horizon = projection.project(geographicLocation, PI_2 - virtualHorizon)
        val alpha = computeHourAngleOfEvent() + topocentric.lon - topos.localTrueSiderealTime.radians
        val delta = Angle.reduce(alpha) * SIDEREAL_MILLIS_PER_RADIAN
        val epochMilliNew = time.epochMilli + delta
        time = time.addMillis(when {
            lookupDirection == LookupDirection.FORWARD && epochMilliNew < startTime.epochMilli -> delta + MILLIS_PER_SIDEREAL_DAY
            lookupDirection == LookupDirection.BACKWARD && epochMilliNew > startTime.epochMilli -> delta - MILLIS_PER_SIDEREAL_DAY
            else -> delta
        }.toLong())
        return !java.lang.Double.isNaN(alpha) && !java.lang.Double.isInfinite(alpha)
    }

    private fun computeHourAngleOfEvent(): Double {
        val dist = horizon.distFromOrigin
        val rHor = horizon.r
        val rOrb = orbitRadius()
        return when (mode) {
            EventType.RISE -> -Math.acos((rHor * rHor - dist * dist - rOrb * rOrb) / (2.0 * dist * rOrb))
            EventType.SET -> Math.acos((rHor * rHor - dist * dist - rOrb * rOrb) / (2.0 * dist * rOrb))
            EventType.TRANSIT -> 0.0
        }
    }

    private fun orbitRadius(): Double {
        val z = Math.sin(if (geographicLocation.lat > 0) topocentric.lat else -topocentric.lat)
        return Math.sqrt((1 + z) / (1 - z))
    }

    protected fun isCrossing() = mode == EventType.TRANSIT || !completelyAboveHorizon() && !completelyBelowHorizon()

    private fun completelyAboveHorizon() =
            Math.abs(geographicLocation.lat + topocentric.lat) >= PI_2 + virtualHorizon

    private fun completelyBelowHorizon() =
            Math.abs(geographicLocation.lat - topocentric.lat) >= PI_2 - virtualHorizon

    companion object {
        val OPTICAL_HORIZON = -Math.toRadians(34.0 / 60)
        const val PI_2 = Math.PI * 0.5
    }
}
